core/src/smartcontracts/isi/triggers/set.rs

```diff
@@ -633,52 +654,96 @@ impl Set {
-                .expect("`Set::time_triggers` doesn't contain required id. This is a bug"),
-            TriggeringEventType::ExecuteTrigger => self
-                .by_call_triggers
-                .remove(id)
-                .map(extract_blob_hash)
-                .expect("`Set::by_call_triggers` doesn't contain required id. This is a bug"),
+        let removed = match event_type {
+            TriggeringEventType::Data => {
+                Self::remove_from(&mut self.original_contracts, &mut self.data_triggers, id)
+            }
+            TriggeringEventType::Pipeline => Self::remove_from(
+                &mut self.original_contracts,
+                &mut self.pipeline_triggers,
+                id,
+            ),
+            TriggeringEventType::Time => {
+                Self::remove_from(&mut self.original_contracts, &mut self.time_triggers, id)
+            }
+            TriggeringEventType::ExecuteTrigger => {
+                Self::remove_from(&mut self.original_contracts, &mut self.by_call_triggers, id)
+            }
         };
 
-        if let Some(blob_hash) = blob_hash {
-            self.original_contracts
-                .remove(&blob_hash)
-                .expect("`Set::original_contracts` doesn't contain required hash. This is a bug");
-        }
+        assert!(
+            removed,
+            "`Set`'s `ids` and typed trigger collections are inconsistent. This is a bug"
+        );
 
         true
     }
 
+    /// Remove trigger from `triggers` and decrease the counter of the original [`WasmSmartContract`].
+    ///
+    /// Note that this function doesn't remove the trigger from [`Set::ids`].
+    ///
+    /// Returns `true` if trigger was removed and `false` otherwise.
+    fn remove_from<F: Filter>(
+        original_contracts: &mut IndexMap<
+            HashOf<WasmSmartContract>,
+            (WasmSmartContract, NonZeroU64),
+        >,
+        triggers: &mut IndexMap<TriggerId, LoadedAction<F>>,
+        trigger_id: &TriggerId,
+    ) -> bool {
+        triggers
+            .remove(trigger_id)
+            .map(|loaded_action| {
+                if let Some(blob_hash) = Self::extract_blob_hash(loaded_action) {
+                    Self::remove_original_trigger(original_contracts, blob_hash);
+                }
+            })
+            .is_some()
+    }
+
+    /// Decrease the counter of the original [`WasmSmartContract`] by `blob_hash`
+    /// or remove it if the counter reaches zero.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `blob_hash` is not in the [`Set::original_contracts`].
+    fn remove_original_trigger(
+        original_contracts: &mut IndexMap<
+            HashOf<WasmSmartContract>,
+            (WasmSmartContract, NonZeroU64),
+        >,
+        blob_hash: HashOf<WasmSmartContract>,
+    ) {
+        (|| {
+            let count = &mut original_contracts.get_mut(&blob_hash)?.1;
+            if let Some(new_count) = NonZeroU64::new(count.get() - 1) {
```

Comment:
Weird that NonZero* types do not have `saturating_sub`, but there is some argument against them: https://internals.rust-lang.org/t/checked-sub-and-saturating-sub-for-nonzero-unsigned-integers/19898/7

Reply:
Do we need `NonZero` here in the first place? It would save us some space in case of `Option<NonZero>`.
But in this case i don't see many benefits in using it.

Reply:
That's... Also fair

While it is true that we have an invariant that the u64 must not be zero, `NonZeroU64`'s hit to ergonomics kinda makes it not worth it to enforce the invariant at the type system level..

Reply:
I think it's more robust than just an integer type, because it introduces invariants which are useful in this scenario. Count should either be > 0 or doesn't exist

Reply:
It's also a documentation for future developers

Reply:
> NonZeroU64's hit to ergonomics kinda makes it not worth it to enforce the invariant at the type system level..

I don't think having just an unsigned integer type would help much, because I will still need to check for zero case

Reply:
I just don't like the `NonZeroU64::new(v.get() - 1)` dance, but no strong feelings, it can stay.